### PR TITLE
[Quality Of Life] Storing named depositories addresses in controller

### PR DESCRIPTION
**Pull request by uxd-vincent**
We talked about clarifying which depositories are part of the main self-balancing set of depositories.

This PR instroduce new fields on the controller, to store and check the allowed depositories selected for rebalancing and to be used in the router.

PR Combo:
- uxd-client: https://github.com/UXDProtocol/uxd-client/pull/43
- uxd-program: https://github.com/UXDProtocol/uxd-program/pull/259
- uxd-cpi: https://github.com/UXDProtocol/uxd-cpi/pull/2

### CODE REVIEW

File: programs/uxd/src/state/controller.rs [hunk 1]
```diff
@@ -81,10 +84,15 @@ pub struct Controller {
     // Total amount of profits collected into the treasury by any depository
     pub profits_total_collected: u128,
 
-    // The configured depository balancing weights
+    // The configured router depositories balancing weights
     pub identity_depository_weight_bps: u16,
-    pub mercurial_vault_depository_0_weight_bps: u16,
```
Comment (acamill) [inline programs/uxd/src/state/controller.rs:86]: Ok this keeps the same memory layout so all good!

File: programs/uxd/tests/integration_tests/api/program_uxd/instructions/process_redeem.rs [hunk 2]
```diff
@@ -109,12 +107,12 @@ pub async fn process_redeem(
...
         identity_depository_collateral_vault,
-        mercurial_vault_depository_0,
-        mercurial_vault_depository_0_vault,
-        mercurial_vault_depository_0_vault_lp_mint: *mercurial_vault_depository_0_vault_lp_mint,
-        mercurial_vault_depository_0_lp_token_vault,
-        mercurial_vault_depository_0_collateral_token_safe,
-        credix_lp_depository_0,
+        mercurial_vault_depository,
+        mercurial_vault_depository_vault,
+        mercurial_vault_depository_vault_lp_mint: *mercurial_vault_depository_vault_lp_mint,
```
Comment (acamill) [inline programs/uxd/tests/integration_tests/api/program_uxd/instructions/process_redeem.rs:112]: <3

**Review by acamill — DISMISSED** [review]
LGTM

File: programs/uxd/src/events.rs [hunk 3]
```diff
@@ -39,8 +39,23 @@ pub struct SetDepositoriesWeightBps {
...
+    pub credix_lp_depository_weight_bps: u16,
+}
+
+/// Event called in [instructions::edit_controller::handler].
+#[event]
+pub struct SetDepositories {
+    /// The controller version.
+    #[index]
+    pub version: u8,
+    /// The controller.
```
Comment (GregoryNEUT) [inline programs/uxd/src/events.rs:52]: nit: remove low value comment

File: programs/uxd/src/instructions/edit_controller.rs [hunk 4]
```diff
@@ -25,14 +26,22 @@ pub struct EditController<'info> {
...
 pub struct EditControllerDepositoriesWeightBps {
     pub identity_depository_weight_bps: u16,
-    pub mercurial_vault_depository_0_weight_bps: u16,
-    pub credix_lp_depository_0_weight_bps: u16,
+    pub mercurial_vault_depository_weight_bps: u16,
+    pub credix_lp_depository_weight_bps: u16,
+}
+
+#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy)]
+pub struct EditControllerDepositories {
```
Comment (GregoryNEUT) [inline programs/uxd/src/instructions/edit_controller.rs:34]: Should we specify somehow here that it's related to `Router`, maybe in the name?

Like that it seems a duplicata of `controller.registered_mercurial_vault_depositories`

File: programs/uxd/src/instructions/edit_controller.rs [hunk 5]
```diff
@@ -25,14 +26,22 @@ pub struct EditController<'info> {
...
+    pub identity_depository: Pubkey,
+    pub mercurial_vault_depository: Pubkey,
+    pub credix_lp_depository: Pubkey,
 }
 
 #[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy)]
 pub struct EditControllerFields {
     pub redeemable_global_supply_cap: Option<u128>,
     pub depositories_weight_bps: Option<EditControllerDepositoriesWeightBps>,
+    pub depositories: Option<EditControllerDepositories>,
```
Comment (GregoryNEUT) [inline programs/uxd/src/instructions/edit_controller.rs:44]: nit: `router_ depositories` ?

File: programs/uxd/src/instructions/edit_controller.rs [hunk 6]
```diff
@@ -41,32 +50,60 @@ pub(crate) fn handler(ctx: Context<EditController>, fields: &EditControllerField
...
+        );
+        msg!(
+            "[edit_controller] mercurial_vault_depository {}",
+            mercurial_vault_depository
+        );
+        msg!(
+            "[edit_controller] credix_lp_depository {}",
+            credix_lp_depository
+        );
+        controller.identity_depository = identity_depository;
```
Comment (GregoryNEUT) [inline programs/uxd/src/instructions/edit_controller.rs:98]: Should we check the provided pubkey is a part of registered controller depositories?

File: programs/uxd/src/instructions/edit_controller.rs (hunk 4 above)
Comment (acamill) [reply programs/uxd/src/instructions/edit_controller.rs:34]: Yes I think that should appear somewhere indeed

File: programs/uxd/src/instructions/edit_controller.rs (hunk 6 above)
Comment (acamill) [reply programs/uxd/src/instructions/edit_controller.rs:98]: As long as we keep them.. yes probably for consistency, although it's a useless check in absolute. We should refactor the concept of registered depository out

File: programs/uxd/src/events.rs (hunk 3 above)
Comment (uxd-vincent) [reply programs/uxd/src/events.rs]: sounds good 👌

File: programs/uxd/src/instructions/edit_controller.rs (hunk 6 above)
Comment (uxd-vincent) [reply programs/uxd/src/instructions/edit_controller.rs]: I think for now we can keep it that way:
- router mint/redeem check for EXACTLY `router_depositories` matches
- regular mint/redeem depository specifics check for "contained in" `registered_depositories`

I think it makes the most sense as they semantically have different purposes, but long term we should remove `registered_depositories` completely IMO, a bit too much for right now in this PR

File: programs/uxd/src/instructions/edit_controller.rs (hunk 5 above)
Comment (uxd-vincent) [reply programs/uxd/src/instructions/edit_controller.rs]: Sounds good king, doing now 👌

File: programs/uxd/src/instructions/edit_controller.rs (hunk 4 above)
Comment (uxd-vincent) [reply programs/uxd/src/instructions/edit_controller.rs]: Sounds good kings, doing now 👌

File: programs/uxd/tests/integration_tests/api/program_uxd/instructions/process_redeem.rs (hunk 2 above)
Comment (uxd-vincent) [reply programs/uxd/tests/integration_tests/api/program_uxd/instructions/process_redeem.rs:112]: 🏩

File: programs/uxd/src/state/controller.rs [hunk 7]
```diff
@@ -81,10 +84,15 @@ pub struct Controller {
...
+    pub mercurial_vault_depository_weight_bps: u16,
+    pub credix_lp_depository_weight_bps: u16,
+
+    // The configured router depositories addresses
+    pub identity_depository: Pubkey,
+    pub mercurial_vault_depository: Pubkey,
+    pub credix_lp_depository: Pubkey,
 
     // For future usage
     pub _reserved: [u8; CONTROLLER_RESERVED_SPACE],
```
[the remark this answers is not in the document]
Comment (uxd-vincent) [reply programs/uxd/src/state/controller.rs:98]: True true, can be done after this release tho IMO, maybe we can do final sets of micro-cleanups in june?

File: programs/uxd/src/state/controller.rs (hunk 1 above)
Comment (uxd-vincent) [reply programs/uxd/src/state/controller.rs:86]: Yes ser, also this part hasnt been deployed yet, so still time to change if we want?

File: programs/uxd/src/instructions/edit_controller.rs (hunk 6 above)
Comment (GregoryNEUT) [reply programs/uxd/src/instructions/edit_controller.rs]: My concern was what happens if you put an address that doesn't exist?

File: programs/uxd/src/state/controller.rs (hunk 1 above)
Comment (acamill) [reply programs/uxd/src/state/controller.rs:86]: No it's cool, most important is to not wreck the memory layout, we are like on a gameboy here so better not loose the offset of each property. 
Eventually we want to refactor and remove the registered depositories but I think that's another beast

File: programs/uxd/src/instructions/edit_controller.rs (hunk 6 above)
Comment (uxd-vincent) [reply programs/uxd/src/instructions/edit_controller.rs]: Yeah I agree.

So the possible cases are:
- 1) wrong router_depositories content -> the router stops working and single mint/redeem keeps working
- 2) wrong registered_depositories content -> everything stops working

I think long term we can get rid of N2

File: programs/uxd/src/state/controller.rs (hunk 1 above)
Comment (uxd-vincent) [reply programs/uxd/src/state/controller.rs:86]: Ok sounds good 👌

**Review by acamill — APPROVED** [review]

**Review by acamill — APPROVED** [review]

**Review by acamill — APPROVED** [review]

**Review by cnek — APPROVED** [review]

**Pull request merged**